src/server/src/main/java/io/cassandrareaper/service/RepairManager.java

```diff
@@ -157,8 +157,11 @@ private void abortAllRunningSegmentsWithNoLeader(Collection<RepairRun> runningRe
         .forEach((repairRun) -> {
           Collection<RepairSegment> runningSegments
               = context.storage.getSegmentsWithState(repairRun.getId(), RepairSegment.State.RUNNING);
+          Collection<RepairSegment> startedSegments
+              = context.storage.getSegmentsWithState(repairRun.getId(), RepairSegment.State.STARTED);
```

Comment:
Instead of grabbing two separate lists, maybe we could modify `getSegmentsWithState()` to allow passing several states we want to filter on at the same time.
I'm concerned with the number of repair scans we're making here, especially on Cassandra. I have another PR in the works that will improve this, and that I hope to merge shortly.
But Reaper is already sending too many requests IMO to the Cassandra backend and I'd like to make it lighter.
wdyt?

Reply:
I had the same idea, but I see no eays way how this could be implemented.

CQL does not allow using `OR` in `WHERE` clauses and the `IN` operator, which would allow to specify multiple values, is only allowed for primary-key columns.

This means that we could only implement the filtering on the client side. Assuming that usually the number of `RUNNING` and `STARTED` segments will be low in comparison to the number of segments in the `NOT_STARTED` and `DONE` states, I think the additional load caused by having to transfer data for all segments might be higher than the load caused by having Cassandra filter the same data twice. As the two queries are executed right after each other, I would expect Cassandra to still have the data in the cache anyway, thus wasting CPU cycles but no I/O resources.

In order to know for sure which solution is better, we of course would have to run some kind of benchmark, but I am not entirely sure whether this effort is justified.

There is another solution, but again I am not sure whether it is worth the effort: We could change the schema of the `repair_run` table and add a new boolean column that tells whether the segment is in the `RUNNING` or `STARTED` state. Then we could filter on this column instead of the `segment_state` column. The disadvantage of this solution is that (apart from requiring a schema change) it would be specifically catered to this use case (finding segments that are running or started).

What do you think? Should we implement either of the two possible solutions?

Reply:
I like the idea of an additional column and you're right, we're using server side filtering with Cassandra >= 3.0 (we're still grabbing all segments and applying the filter on the client side for C* <= 2.2.x) and we can't select both RUNNING and STARTED.
If we come up with some fancy new segment states in the future, this additional column would help us distinguish state groups.